mixer/adapter/svcctrl/expiringLRUCache.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package svcctrl
+
+import (
+	"sync"
+	"time"
+
+	"github.com/hashicorp/golang-lru/simplelru"
+	"github.com/jonboulle/clockwork"
+)
+
+type (
+	cacheValue struct {
+		// insert timestamp
+		timestamp time.Time
+		value     interface{}
+	}
+
+	// A thread safe LRU cache with expiration.
+	expiringLRUCache struct {
+		mutex sync.RWMutex
```

Comment:
you can just have this be:

```
expiringLRUCache struct {
  sync.RWMutex
  ...
}
```

Then you can use it as:
```
    c := newCache(...)
    ...
    c.Lock()
    c.lru.Add(...)
    c.Unlock()
```

However, I'm still not clear on why this needs a mutex, given that the underlying data structure is thread-safe.

Reply:
It has a thread-safe version built on top of the non-thread-safe version, which I used.  The non-thread safe version is easier to add time expiration with high performance.